sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/PresenceConsolidation.java

```diff
@@ -0,0 +1,284 @@
+package org.sakaiproject.sitestats.impl;
+
+import java.lang.reflect.InvocationTargetException;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+import org.apache.commons.beanutils.BeanUtils;
+import org.sakaiproject.sitestats.api.presence.Presence;
+
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class PresenceConsolidation {
+
+
+    private static final Comparator<PresenceRecord> PRESENCE_RECORDS_ORDER = Comparator.comparing(Presence::getBegin);
+
+    private List<PresenceRecord> records;
+
+
+    public PresenceConsolidation() {
+        records = new ArrayList<>();
+    }
+
+
+    public boolean add(@NonNull Presence presence) {
+        if (!presence.isComplete()) {
+            throw new IllegalArgumentException("Can not add incomplete record to consolidation");
+        }
+
+        if (records.contains(presence)) {
+            // Record is already present in list, leave unchanged
+            return false;
+        }
+
+        PresenceRecord firstOverlappingRecord = null;
+        PresenceRecord lastOverlappingRecord = null;
+
+        // Assign first and last overlapping records
+        for (PresenceRecord currentRecord : records) {
+            if (currentRecord.overlapsWith(presence)) {
+                if (firstOverlappingRecord == null) {
+                    firstOverlappingRecord = currentRecord;
+                }
+
+                lastOverlappingRecord = currentRecord;
+            } else if (lastOverlappingRecord != null) {
+                break;
+            }
+        }
+
+        // If we have no intersections, just add the new record
+        if (firstOverlappingRecord == null && lastOverlappingRecord == null) {
+            records.add(presenceToRecord(presence));
+            return true;
+        }
+
+        // If there is only one intersection, extend the existing record
+        if (firstOverlappingRecord != null && firstOverlappingRecord == lastOverlappingRecord) {
```

Comment:
is == appropriate here? Or do you need to compare?

Reply:
Yes, this is checking if the first and last overlapping record is actually the same object. It should be impossible that there are multiple equal records within the consolidation, that's the second check in this method.